### PR TITLE
currentlyTicking Solution

### PR DESCRIPTION
**Pull request by ZeroMemes**
No more Helper methods.
👋 
``player()``
``world()``
``playerController()``

### CODE REVIEW

File: src/api/java/baritone/api/utils/IPlayerContext.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    EntityPlayerSP player();
+
+    PlayerControllerMP playerController();
+
+    World world();
+
+    IWorldData worldData();
+
+    BetterBlockPos playerFeet();
```
Comment (leijurv) [inline src/api/java/baritone/api/utils/IPlayerContext.java]: Can the +0.1251 meme be here? Can just do world().getBlockState, no?

File: src/api/java/baritone/api/utils/input/Input.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    private static final Map<KeyBinding, Input> bindToInputMap = new HashMap<>();
+
+    /**
+     * The actual game {@link KeyBinding} being forced.
+     */
+    private final KeyBinding keyBinding;
+
+    Input(Function<GameSettings, KeyBinding> keyBindingMapper) {
+        this.keyBinding = keyBindingMapper.apply(Minecraft.getMinecraft().gameSettings);
```
Comment (leijurv) [inline src/api/java/baritone/api/utils/input/Input.java]: I don't understand why the lambdas... Can't it just be as before, mc.gameSettings.keyBindX?

File: src/main/java/baritone/Baritone.java [hunk 3]
```diff
@@ -162,6 +165,11 @@ public GetToBlockProcess getGetToBlockProcess() {  // Iffy
         return getToBlockProcess;
     }
 
+    @Override
+    public IPlayerContext getPlayerContext() {
+        return LocalPlayerContext.INSTANCE;
```
Comment (leijurv) [inline src/main/java/baritone/Baritone.java]: Shouldn't this be baritone instance specific?

File: src/main/java/baritone/behavior/MemoryBehavior.java [hunk 4]
```diff
@@ -120,24 +120,24 @@ public synchronized void onReceivePacket(PacketEvent event) {
...
     }
 
     private void updateInventory() {
-        getInventoryFromWindow(player().openContainer.windowId).ifPresent(inventory -> {
+        getInventoryFromWindow(ctx.player().openContainer.windowId).ifPresent(inventory -> {
             inventory.items.clear();
-            inventory.items.addAll(player().openContainer.getInventory().subList(0, inventory.size));
+            inventory.items.addAll(ctx.player().openContainer.getInventory().subList(0, inventory.size));
         });
     }
```
Comment (leijurv) [inline src/main/java/baritone/behavior/MemoryBehavior.java]: General on MemoryBehavior: could we instead hook into where the packets are processed on the main thread, instead of where netty receives them? So that we avoid any possible concurrent modification exceptions?

File: src/main/java/baritone/behavior/PathingBehavior.java [hunk 5]
```diff
@@ -421,7 +421,7 @@ private void findPathInNewThread(final BlockPos start, final boolean talkAboutIt
                 }
 
                 return result;
-            }).map(PathExecutor::new);
+            }).map(p -> new PathExecutor(this, p));
```
Comment (leijurv) [inline src/main/java/baritone/behavior/PathingBehavior.java]: this is SO sad

File: src/main/java/baritone/cache/CachedWorld.java [hunk 6]
```diff
@@ -106,10 +106,10 @@ public final boolean isCached(int blockX, int blockZ) {
     }
 
     @Override
-    public final LinkedList<BlockPos> getLocationsOf(String block, int maximum, int maxRegionDistanceSq) {
+    public final LinkedList<BlockPos> getLocationsOf(String block, int maximum, int centerX, int centerZ, int maxRegionDistanceSq) {
         LinkedList<BlockPos> res = new LinkedList<>();
-        int playerRegionX = playerFeet().getX() >> 9;
-        int playerRegionZ = playerFeet().getZ() >> 9;
+        int centerRegionX = centerX >> 9;
+        int centerRegionZ = centerZ >> 9;
```
Comment (leijurv) [inline src/main/java/baritone/cache/CachedWorld.java]: Can you remove the `implements Helper` for all the classes that no longer use it? (like this one I think)

File: src/main/java/baritone/pathing/movement/CalculationContext.java [hunk 7]
```diff
@@ -85,6 +87,10 @@ public CalculationContext() {
         this.worldBorder = new BetterWorldBorder(world.getWorldBorder());
     }
 
+    public final IBaritone getBaritone() {
+        return this.baritone;
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/CalculationContext.java]: REEEEEEEEEEEee nothing else in this class uses this.

when i come into your house (the `cache` package), i follow your rules, i take off my shoes, i use `this.`

when you come into my house, i expect you to follow my rules (keep your shoes on at all times) and don't use `this.`

File: src/main/java/baritone/pathing/movement/Movement.java [hunk 8]
```diff
@@ -160,11 +163,11 @@ protected boolean prepared(MovementState state) {
                 //i'm doing it anyway
                 //i dont care if theres snow in the way!!!!!!!
                 //you dont own me!!!!
-                state.setTarget(new MovementState.MovementTarget(RotationUtils.calcRotationFromVec3d(player().getPositionEyes(1.0F),
+                state.setTarget(new MovementState.MovementTarget(RotationUtils.calcRotationFromVec3d(ctx.player().getPositionEyes(1.0F),
                         VecUtils.getBlockPosCenter(blockPos)), true)
                 );
                 // don't check selectedblock on this one, this is a fallback when we can't see any face directly, it's intended to be breaking the "incorrect" block
-                state.setInput(InputOverrideHandler.Input.CLICK_LEFT, true);
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/Movement.java]: I do like these, these are epic

File: src/main/java/baritone/pathing/movement/Movement.java [hunk 9]
```diff
@@ -108,32 +111,32 @@ public double calculateCostWithoutCaching() {
...
         }
-        if (player().isEntityInsideOpaqueBlock()) {
+        if (ctx.player().isEntityInsideOpaqueBlock()) {
             currentState.setInput(Input.CLICK_LEFT, true);
         }
 
         // If the movement target has to force the new rotations, or we aren't using silent move, then force the rotations
         currentState.getTarget().getRotation().ifPresent(rotation ->
-                Baritone.INSTANCE.getLookBehavior().updateTarget(
+                baritone.getLookBehavior().updateTarget(
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/Movement.java]: Remind me why baritone can't just be a parameter passed to update()??

File: src/main/java/baritone/pathing/movement/MovementHelper.java [hunk 10]
```diff
@@ -44,33 +44,27 @@
...
         Block b = state.getBlock();
         return b == Blocks.ICE // ice becomes water, and water can mess up the path
                 || b instanceof BlockSilverfish // obvious reasons
                 // call context.get directly with x,y,z. no need to make 5 new BlockPos for no reason
-                || context.get(x, y + 1, z).getBlock() instanceof BlockLiquid//don't break anything touching liquid on any side
-                || context.get(x + 1, y, z).getBlock() instanceof BlockLiquid
-                || context.get(x - 1, y, z).getBlock() instanceof BlockLiquid
-                || context.get(x, y, z + 1).getBlock() instanceof BlockLiquid
-                || context.get(x, y, z - 1).getBlock() instanceof BlockLiquid;
+                || bsi.get0(x, y + 1, z).getBlock() instanceof BlockLiquid//don't break anything touching liquid on any side
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/MovementHelper.java]: @HowardStark kills self

File: src/main/java/baritone/pathing/movement/MovementHelper.java [hunk 11]
```diff
@@ -111,7 +105,7 @@ static boolean canWalkThrough(CalculationContext context, int x, int y, int z, I
             if (Baritone.settings().assumeWalkOnWater.get()) {
                 return false;
             }
-            IBlockState up = context.get(x, y + 1, z);
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/MovementHelper.java]: I don't understand, why these changes?

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java [hunk 12]
```diff
@@ -44,23 +45,23 @@
 
 public class MovementParkour extends Movement {
 
-    private static final EnumFacing[] HORIZONTALS_BUT_ALSO_DOWN____SO_EVERY_DIRECTION_EXCEPT_UP = {EnumFacing.NORTH, EnumFacing.SOUTH, EnumFacing.EAST, EnumFacing.WEST, EnumFacing.DOWN};
+    private static final EnumFacing[] HORIZONTAL_AND_DOWN = { EnumFacing.NORTH, EnumFacing.SOUTH, EnumFacing.EAST, EnumFacing.WEST, EnumFacing.DOWN };
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: Not going to merge because of this change.

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java [hunk 13]
```diff
@@ -110,7 +111,7 @@ public static void cost(CalculationContext context, int x, int y, int z, EnumFac
                     return;
                 }
             }
-            if (MovementHelper.canWalkOn(context, x + xDiff * i, y - 1, z + zDiff * i)) {
+            if (MovementHelper.canWalkOn(context.bsi(), x + xDiff * i, y - 1, z + zDiff * i)) {
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: This is just icky... why?

File: src/main/java/baritone/pathing/movement/movements/MovementTraverse.java [hunk 14]
```diff
@@ -152,86 +153,86 @@ public MovementState updateState(MovementState state) {
...
-        state.setInput(InputOverrideHandler.Input.SNEAK, false);
+        state.setInput(Input.SNEAK, false);
 
-        Block fd = BlockStateInterface.get(src.down()).getBlock();
+        Block fd = BlockStateInterface.get(ctx, src.down()).getBlock();
         boolean ladder = fd instanceof BlockLadder || fd instanceof BlockVine;
-        IBlockState pb0 = BlockStateInterface.get(positionsToBreak[0]);
-        IBlockState pb1 = BlockStateInterface.get(positionsToBreak[1]);
+        IBlockState pb0 = BlockStateInterface.get(ctx, positionsToBreak[0]);
+        IBlockState pb1 = BlockStateInterface.get(ctx, positionsToBreak[1]);
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/movements/MovementTraverse.java]: I do not understand, why do we sometimes do ctx.get, ctx.bsi().get0...
wait 
i just remembered that ctx can be either IPlayerContext or CalculationContext

vomit vomit vomit!

File: src/main/java/baritone/process/MineProcess.java [hunk 15]
```diff
@@ -90,7 +88,7 @@ public PathingCommand onTick(boolean calcFailed, boolean isSafeToCancel) {
         int mineGoalUpdateInterval = Baritone.settings().mineGoalUpdateInterval.get();
         if (mineGoalUpdateInterval != 0 && tickCount++ % mineGoalUpdateInterval == 0) { // big brain
             List<BlockPos> curr = new ArrayList<>(knownOreLocations);
-            baritone.getExecutor().execute(() -> rescan(curr));
+            Baritone.getExecutor().execute(() -> rescan(curr));
```
Comment (leijurv) [inline src/main/java/baritone/process/MineProcess.java]: agree

File: src/main/java/baritone/process/MineProcess.java [hunk 16]
```diff
@@ -170,26 +168,15 @@ private void rescan(List<BlockPos> already) {
...
-            if (downwardGoal) {
-                return new GoalBlock(loc.down());
-            } else {
-                return new GoalTwoBlocks(loc);
-            }
-        }
+        // Here, BlockStateInterface is used because the position may be in a cached chunk (the targeted block is one that is kept track of)
+        boolean upwardGoal = locs.contains(loc.up()) || (Baritone.settings().internalMiningAirException.get() && BlockStateInterface.getBlock(ctx, loc.up()) == Blocks.AIR);
+        boolean downwardGoal = locs.contains(loc.down()) || (Baritone.settings().internalMiningAirException.get() && BlockStateInterface.getBlock(ctx, loc.down()) == Blocks.AIR);
+        return upwardGoal == downwardGoal ? new GoalTwoBlocks(loc) : upwardGoal ? new GoalBlock(loc) : new GoalBlock(loc.down());
```
Comment (leijurv) [inline src/main/java/baritone/process/MineProcess.java]: hell yeah

File: src/main/java/baritone/utils/BlockStateInterface.java [hunk 17]
```diff
@@ -38,28 +40,31 @@
...
-    public static Block getBlock(BlockPos pos) { // won't be called from the pathing thread because the pathing thread doesn't make a single blockpos pog
-        return get(pos).getBlock();
+    public static Block getBlock(IPlayerContext ctx, BlockPos pos) { // won't be called from the pathing thread because the pathing thread doesn't make a single blockpos pog
+        return get(ctx, pos).getBlock();
     }
 
-    public static IBlockState get(BlockPos pos) {
-        return new CalculationContext().get(pos); // immense iq
+    public static IBlockState get(IPlayerContext ctx, BlockPos pos) {
+        return new BlockStateInterface(ctx).get0(pos.getX(), pos.getY(), pos.getZ()); // immense iq
```
Comment (leijurv) [inline src/main/java/baritone/utils/BlockStateInterface.java]: Okay I'll give you that it's better to just create a BlockStateInterface not a CalculationContext but still, not the best. Idk.

File: src/main/java/baritone/utils/player/AbstractPlayerContext.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+import baritone.api.utils.BetterBlockPos;
+import baritone.api.utils.IPlayerContext;
+import net.minecraft.block.BlockSlab;
+
+/**
+ * @author Brady
+ * @since 11/12/2018
+ */
+public abstract class AbstractPlayerContext implements IPlayerContext {
```
Comment (leijurv) [inline src/main/java/baritone/utils/player/AbstractPlayerContext.java]: dont get why this intermediary class needs to exist tbh

File: src/main/java/baritone/utils/player/LocalPlayerContext.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+import net.minecraft.client.multiplayer.PlayerControllerMP;
+import net.minecraft.world.World;
+
+/**
+ * Implementation of {@link IPlayerContext} that provides information about the local player.
+ *
+ * @author Brady
+ * @since 11/12/2018
+ */
+public final class LocalPlayerContext extends AbstractPlayerContext {
```
Comment (leijurv) [inline src/main/java/baritone/utils/player/LocalPlayerContext.java]: So for the "primary" Baritone instance, the one that's created by default and refers to the current player, this is the context it uses? Ok, I'm fine with that.

Comment (ZeroMemes) [reply src/main/java/baritone/utils/player/LocalPlayerContext.java]: Yeah this is an implementation specifically for that. ``bot-system`` will use another one.

File: src/main/java/baritone/utils/player/AbstractPlayerContext.java (hunk 18 above)
Comment (ZeroMemes) [reply src/main/java/baritone/utils/player/AbstractPlayerContext.java]: It's to implement ``playerFeet()`` just once, so it'll apply to ``LocalPlayerContext`` and whatever the ``BotPlayerContext`` will be.

File: src/main/java/baritone/utils/BlockStateInterface.java (hunk 17 above)
Comment (ZeroMemes) [reply src/main/java/baritone/utils/BlockStateInterface.java]: nOt the best? we're technically I'm removing like 3 method calls.

<details>
Owned with facts and logic.
</details>

Comment (leijurv) [reply src/main/java/baritone/utils/BlockStateInterface.java]: Yes it's better but it's not ideal, that's what i said.

<details>
rekt
</details>

File: src/main/java/baritone/utils/player/AbstractPlayerContext.java (hunk 18 above)
Comment (leijurv) [reply src/main/java/baritone/utils/player/AbstractPlayerContext.java]: yeah i'm saying define it as a default in `IPlayerContext`

Comment (ZeroMemes) [reply src/main/java/baritone/utils/player/AbstractPlayerContext.java]: Oh I'm dumb lol!

File: src/main/java/baritone/pathing/movement/Movement.java (hunk 9 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/Movement.java]: There would be a lot of other methods to add a parameter to and Movement is Baritone-instance based anyways

File: src/main/java/baritone/pathing/movement/Movement.java (hunk 8 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/Movement.java]: Wait what?

File: src/main/java/baritone/pathing/movement/Movement.java (hunk 9 above)
Comment (leijurv) [reply src/main/java/baritone/pathing/movement/Movement.java]: Like what? In both scenarios, the code within updateState has a reference to the Baritone?

File: src/main/java/baritone/utils/BlockStateInterface.java (hunk 17 above)
Comment (ZeroMemes) [reply src/main/java/baritone/utils/BlockStateInterface.java]: FUU how else could I improve this?

File: src/main/java/baritone/pathing/movement/Movement.java (hunk 8 above)
Comment (leijurv) [reply src/main/java/baritone/pathing/movement/Movement.java]: I like the removal of InputOverrideHandler., thumbs up

File: src/main/java/baritone/utils/BlockStateInterface.java (hunk 17 above)
Comment (leijurv) [reply src/main/java/baritone/utils/BlockStateInterface.java]: Eh it's good enough IMO.

File: src/main/java/baritone/cache/CachedWorld.java (hunk 6 above)
Comment (ZeroMemes) [reply src/main/java/baritone/cache/CachedWorld.java]: ``CachedWorld`` actually does use it in other places, but removed it in all the other places where it's not used.

File: src/api/java/baritone/api/utils/input/Input.java (hunk 2 above)
Comment (ZeroMemes) [reply src/api/java/baritone/api/utils/input/Input.java]: Enums can't refer to static fields like in their dummy constructors and I didn't want to create another interface just to refer to ``mc.[...]``

Comment (ZeroMemes) [reply src/api/java/baritone/api/utils/input/Input.java]: Like I would've had to mirror how ``Helper`` had a ``static`` ``Minecraft`` field in a new interface that would be implemented by ``Input``

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java (hunk 12 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: Fixed lol

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java (hunk 13 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: Part of cost calculation, wanted to reduce amount of calls while keeping canWalkOn usable by updateState, which doesn't have a calc context

File: src/api/java/baritone/api/utils/input/Input.java (hunk 2 above)
Comment (leijurv) [reply src/api/java/baritone/api/utils/input/Input.java]: Ok that is quite weird. So is Helper deleted? I guess this is okay, just add a comment explaining why 😘

Comment (ZeroMemes) [reply src/api/java/baritone/api/utils/input/Input.java]: Not deleted, this class is in ``api``.

File: src/main/java/baritone/behavior/MemoryBehavior.java (hunk 4 above)
Comment (leijurv) [reply src/main/java/baritone/behavior/MemoryBehavior.java]: Hmmmm. Are you sure? 0x22 was getting the h00000 error in MemoryBehavior, which is caused when it's called from a different thread. Psure the current hook is on a different thread. I'm not referring to context, just to possible ConcurrentModificationExceptions in the remembered chest contents.

File: src/main/java/baritone/pathing/movement/Movement.java (hunk 9 above)
Comment (leijurv) [reply src/main/java/baritone/pathing/movement/Movement.java]: fineeeeeeee

File: src/main/java/baritone/pathing/movement/MovementHelper.java (hunk 11 above)
Comment (leijurv) [reply src/main/java/baritone/pathing/movement/MovementHelper.java]: ok now you're speaking my language with words like "performance" referring to cost calculation. hell yeah now i agree. it make baritone like sanik.

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java (hunk 13 above)
Comment (leijurv) [reply src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: Fair

File: src/api/java/baritone/api/utils/input/Input.java (hunk 2 above)
Comment (leijurv) [reply src/api/java/baritone/api/utils/input/Input.java]: wow kank okay, pls explain in comment

File: src/main/java/baritone/Baritone.java [hunk 20]
```diff
@@ -192,6 +200,11 @@ public WorldProvider getWorldProvider() {
         return worldProvider;
     }
 
+    /**
+     * TODO-yeet This shouldn't be baritone-instance specific
```
Comment (leijurv) [inline src/main/java/baritone/Baritone.java]: Why's that? WorldScanner take in a world, right?

File: src/main/java/baritone/behavior/MemoryBehavior.java (hunk 4 above)
Comment (ZeroMemes) [reply src/main/java/baritone/behavior/MemoryBehavior.java]: Yes, by packet processing I meant the handling of "what the packet does". The packet hook that we use in MemoryBehavior is called from the network handler thread, before the packet is actually processed in the main thread. The code in the game is really cancer, but basically it makes an attempt to process packets in the handling thread, and then an exception is thrown terminating the thread after it has been queued to be processed in the main thread.

File: src/main/java/baritone/Baritone.java (hunk 20 above)
Comment (ZeroMemes) [reply src/main/java/baritone/Baritone.java]: No, it takes an ``IPlayerContext`` because it refers to the player position to determine the scan radius, and the world to scan.

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java [hunk 21]
```diff
@@ -44,23 +45,23 @@
 
 public class MovementParkour extends Movement {
 
-    private static final EnumFacing[] HORIZONTALS_BUT_ALSO_DOWN____SO_EVERY_DIRECTION_EXCEPT_UP = {EnumFacing.NORTH, EnumFacing.SOUTH, EnumFacing.EAST, EnumFacing.WEST, EnumFacing.DOWN};
+    private static final EnumFacing[] HORIZONTALS_BUT_ALSO_DOWN____SO_EVERY_DIRECTION_EXCEPT_UP  = { EnumFacing.NORTH, EnumFacing.SOUTH, EnumFacing.EAST, EnumFacing.WEST, EnumFacing.DOWN };
```
Comment (leijurv) [inline src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: screen shot 2018-11-13 at 2 47 01 pm

File: src/main/java/baritone/utils/Helper.java [hunk 22]
```diff
@@ -51,6 +44,7 @@
 
     Minecraft mc = Minecraft.getMinecraft();
 
+    /*
```
Comment (leijurv) [inline src/main/java/baritone/utils/Helper.java]: =O

File: src/main/java/baritone/behavior/MemoryBehavior.java (hunk 4 above)
Comment (leijurv) [reply src/main/java/baritone/behavior/MemoryBehavior.java]: OK i guess we just need to be really careful synchronizing this

File: src/api/java/baritone/api/utils/input/Input.java (hunk 2 above)
Comment (ZeroMemes) [reply src/api/java/baritone/api/utils/input/Input.java]: k

File: src/main/java/baritone/utils/Helper.java (hunk 22 above)
Comment (ZeroMemes) [reply src/main/java/baritone/utils/Helper.java]: LOL yeah should probably just remove it entirely now.

File: src/main/java/baritone/behavior/MemoryBehavior.java (hunk 4 above)
Comment (ZeroMemes) [reply src/main/java/baritone/behavior/MemoryBehavior.java]: Yep.

File: src/main/java/baritone/pathing/movement/MovementHelper.java (hunk 11 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/MovementHelper.java]: awesome style

File: src/main/java/baritone/pathing/movement/movements/MovementParkour.java (hunk 21 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/movements/MovementParkour.java]: LOL fixed.

The second space is because I don't want autism arrays.

File: src/main/java/baritone/pathing/movement/movements/MovementTraverse.java (hunk 14 above)
Comment (ZeroMemes) [reply src/main/java/baritone/pathing/movement/movements/MovementTraverse.java]: Yeahhh. All usages of ``ctx`` should be ``IPlayerContext``, so that shoullddddd be fine.

File: src/main/java/baritone/Baritone.java (hunk 20 above)
Comment (leijurv) [reply src/main/java/baritone/Baritone.java]: Oh whoops I'm blind, I read your comment as saying that it should be baritone instance specific. I agree that it shouldn't be.

Comment (ZeroMemes) [reply src/main/java/baritone/Baritone.java]: Awesome
